### ISSUE TITLE
DDPG/TD3 policy network output clip

### ISSUE THREAD

**Issue by huxiao09** [vwxyzjn/cleanrl#196]
## Problem Description
It seems that the output of policy (actor) network in DDPG/TD3 has been clipped to fit the action range boundaries. But in Fujimoto's DDPG or TD3 code[1] and some other implementations, the **tanh** function is add to the output layer in actor networks to fit the action range, and they don't use **clip**. Have u ever tried the **tanh** implementation? Intuitively the **tanh** implementation could be better, because any action beyond the boundary is considered the same in the **clip** implementation.

```
actions = actor(torch.Tensor(obs).to(device))
                actions = np.array(
                    [
                        (
                            actions.tolist()[0]
                            + np.random.normal(0, max_action * args.exploration_noise, size=envs.single_action_space.shape[0])
                        ).clip(envs.single_action_space.low, envs.single_action_space.high)
                    ]
                )
```

[1] https://github.com/sfujim/TD3

**Comment by vwxyzjn** [comment]
Thanks for raising this issue. I  think you are referring to https://github.com/sfujim/TD3/blob/385b33ac7de4767bab17eb02ade4a268d3e4e24f/TD3.py#L28? Yeah this is a bit of a problem. That said, it's fine in MuJoCo because `max_action` is just 1.

**Comment by huxiao09** [comment]
Yep. I still highly suggest you adopt Fujimoto's implementations, because in some MuJoCo environments (like Humanoid-v2) and some other environments the **max_action** may not be 1.

And I also highly suggest you add the implementation of **pytorch network model saving** to each of your single file, and then write a separate single file that loads the network model and tests it in the corresponding environment.

Thanks for your reply and your nice cleanRL code : )

**Comment by vwxyzjn** [comment]
Thanks for the suggestion! Fixing the `max_action` issue with the actor makes sense to me.

Model saving on the other hand is a bit more nuanced. For example,  [ppo_continuous_action.py](https://github.com/vwxyzjn/cleanrl/blob/master/cleanrl/ppo_continuous_action.py) technically also needs to save the running mean and standard deviations stored in the environment wrappers. Because of these complications, we have decided not doing model saving to keep the core implementation as simple as possible. Additionally, most people save the models only to load them later to see what the agent is actually doing. For this use case, we already included videos of the agents playing the game in the docs (e.g. [link](https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/Pettingzoo-s-Multi-agent-Atari-CleanRL-s-PPO--VmlldzoyMDkxNTE5), further eliminating the use case of saving models.

**Comment by vwxyzjn** [comment]
@dosssman This is indeed a much larger problem - I just tested it out and there is a number of environments that do not have 1 as the default `max_action`.

```
Testing environment: Ant-v2 Observation space: Box(-inf, inf, (111,), float64) Action space: Box(-1.0, 1.0, (8,), float32)
Testing environment: HalfCheetah-v2 Observation space: Box(-inf, inf, (17,), float64) Action space: Box(-1.0, 1.0, (6,), float32)
Testing environment: Hopper-v2 Observation space: Box(-inf, inf, (11,), float64) Action space: Box(-1.0, 1.0, (3,), float32)
Testing environment: Humanoid-v2 Observation space: Box(-inf, inf, (376,), float64) Action space: Box(-0.4, 0.4, (17,), float32)
Testing environment: InvertedDoublePendulum-v2 Observation space: Box(-inf, inf, (11,), float64) Action space: Box(-1.0, 1.0, (1,), float32)
Testing environment: InvertedPendulum-v2 Observation space: Box(-inf, inf, (4,), float64) Action space: Box(-3.0, 3.0, (1,), float32)
Testing environment: Pusher-v2 Observation space: Box(-inf, inf, (23,), float64) Action space: Box(-2.0, 2.0, (7,), float32)
Testing environment: Reacher-v2 Observation space: Box(-inf, inf, (11,), float64) Action space: Box(-1.0, 1.0, (2,), float32)
Testing environment: Swimmer-v2 Observation space: Box(-inf, inf, (8,), float64) Action space: Box(-1.0, 1.0, (2,), float32)
Testing environment: Walker2d-v2 Observation space: Box(-inf, inf, (17,), float64) Action space: Box(-1.0, 1.0, (6,), float32)
```

@huxiao09 would you be interested in submitting a fix for this and going through the process of re-running benchmarks? Otherwise, no worries - we will take care of this.

**Comment by huxiao09** [comment]
Sure, it's my pleasure. But I am busy this week. I'm afraid I can only try my best to finish it in two weeks. Is that too late?

**Comment by dosssman** [comment]
Thanks for the heads up. Will look into this.

**Comment by vwxyzjn** [comment]
Two weeks is actually a great timeline. Thank you @huxiao09 for your interest in working on this!

The process would go like this
1. file a PR to make the changes with the `max_action`
2. we agree on the changes
3. have you join our wandb team `openrlbenchmark` that tracks all of our benchmark experiments
4. run the following benchmark scripts
https://github.com/vwxyzjn/cleanrl/blob/6387191dbee74927b2872b2eb1759c72361d806f/benchmark/ddpg.sh#L1-L7
https://github.com/vwxyzjn/cleanrl/blob/6387191dbee74927b2872b2eb1759c72361d806f/benchmark/td3.sh#L1-L7

Btw would this impact `sac_continouous_action.py` @dosssman? I see in the SAC file there is no clip action like done in PPO:

https://github.com/vwxyzjn/cleanrl/blob/6387191dbee74927b2872b2eb1759c72361d806f/cleanrl/ppo_continuous_action.py#L89

**Comment by dosssman** [comment]
SAC should be unaffected as the actions are scale to the min, max actions before being returned to the algorithm.

Android 版 Outlook<https://aka.ms/ghei36> の入手
________________________________

**Comment by huxiao09** [comment]
ok, get it :)

**Comment by huxiao09** [comment]
Yes, I have already had an account: https://wandb.ai/huxiao .  But it seems that discord can't be open in my region. I'll find how to solve this problem tomorrow.

**Comment by dosssman** [comment]
Preliminary fix tracked [here](https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/MuJoCo-CleanRL-s-TD3-Action-Bound-Fix-Check--VmlldzoyMjAwMjM5).

**Comment by vwxyzjn** [comment]
I changed the run set name to as follows (customize the advanced legend to be `${runsetName}`. 

<img width="1285" alt="image" src="https://user-images.githubusercontent.com/5555347/175541483-9a1d221f-b6ef-492d-82fc-889c5b9afe1a.png">

Could you do the following
- [ ] remove the new `td3_continuous_action_bound_fix.py` experiments with `HalfCheetah-v2` (since its action space's bound is -1, +1)
- [ ] move the old `td3_continuous_action.py` experiments with `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2` to the `openrlbenchmark/cleanrl-cache` project.
- [ ] Add the new `td3_continuous_action_bound_fix.py` experiments with `Humanoid-v2`, `Pusher-v2` experiments to https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/MuJoCo-CleanRL-s-TD3--VmlldzoxNjk4Mzk5 (in particular customize the runset name to be `CleanRL's td3_continuous_action.py` and set the advanced legend to be `${runsetName}`)
- [ ] update the benchmark file https://github.com/vwxyzjn/cleanrl/blob/master/benchmark/td3.sh to include `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2`. 
- [ ] Add a section in the documentation updating charts for `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2` and explaining the use of action scale and bias.

Thank you so much!

**Comment by dosssman** [comment]
Thank you very much for the feedback.
Will fix and add re-runs just to be sure.

**Comment by vwxyzjn** [comment]
Thank you Rousslan. Please remove the experiments or move them to cleanrl-cache project - the stale experiments should not remain in the openrlbenchmark/cleanrl​.
________________________________

**Comment by dosssman** [comment]
> I changed the run set name to as follows (customize the advanced legend to be `${runsetName}`.
> 
> <img alt="image" width="1285" src="https://user-images.githubusercontent.com/5555347/175541483-9a1d221f-b6ef-492d-82fc-889c5b9afe1a.png">
> 
> Could you do the following
> 
> * [ ]  remove the new `td3_continuous_action_bound_fix.py` experiments with `HalfCheetah-v2` (since its action space's bound is -1, +1)
> * [ ]  move the old `td3_continuous_action.py` experiments with `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2` to the `openrlbenchmark/cleanrl-cache` project.
> * [ ]  Add the new `td3_continuous_action_bound_fix.py` experiments with `Humanoid-v2`, `Pusher-v2` experiments to https://wandb.ai/openrlbenchmark/openrlbenchmark/reports/MuJoCo-CleanRL-s-TD3--VmlldzoxNjk4Mzk5 (in particular customize the runset name to be `CleanRL's td3_continuous_action.py` and set the advanced legend to be `${runsetName}`)
> * [ ]  update the benchmark file https://github.com/vwxyzjn/cleanrl/blob/master/benchmark/td3.sh to include `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2`.
> * [ ]  Add a section in the documentation updating charts for `Humanoid-v2`, `Pusher-v2`, and `InvertedPendulum-v2` and explaining the use of action scale and bias.
> 
> Thank you so much!

No worries. Will do.
I tweak the last two lines @huxiao09 mentioned and re-runs the experiment to be sure there is no side-effect, then clean up the benchmark project structure and update the docs.

Should get it done by the next Tuesday or Wednesday.
Thanks.

**Comment by dosssman** [comment]
Let's just keep those old experiments to compare with the last variant.
Once the latest one is validated, I can clean up all the previous variant and re-run "td3_continuous_action.py" to prevent future confusions.

**Comment by dosssman** [comment]
re-runs are looking good so far, not much difference with the previous iterations.
@vwxyzjn It seems I will have to trouble you regarding the archival / removal of the olds td3 and ddpg runs. Wandb mentions missing permissions when I tried to either move the runs to the cleanrl-cache project or just delete them, before queuing up clean "td3_continuous_actions.py" and "ddpg_continuous_actions.py" runs.
